includes/Core/Util/Migration_1_128_0.php

```diff
@@ -0,0 +1,187 @@
+
+	/**
+	 * Context instance.
+	 *
+	 * @since 1.128.0
+	 * @var Context
+	 */
+	protected $context;
+
+	/**
+	 * Options instance.
+	 *
+	 * @since 1.128.0
+	 * @var Options
+	 */
+	protected $options;
+
+	/**
+	 * Analytics_Settings instance.
+	 *
+	 * @since 1.128.0
+	 * @var Analytics_Settings
+	 */
+	protected $analytics_settings;
+
+	/**
+	 * Ads_Settings instance.
+	 *
+	 * @since 1.128.0
+	 * @var Ads_Settings
+	 */
+	protected $ads_settings;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 1.128.0
+	 *
+	 * @param Context $context Plugin context instance.
+	 * @param Options $options Optional. Options instance.
+	 */
+	public function __construct(
+		Context $context,
+		Options $options = null
+	) {
+		$this->context            = $context;
+		$this->options            = $options ?: new Options( $context );
+		$this->analytics_settings = new Analytics_Settings( $this->options );
+		$this->ads_settings       = new Ads_Settings( $this->options );
+	}
+
+	/**
+	 * Registers hooks.
+	 *
+	 * @since 1.128.0
+	 */
+	public function register() {
+		add_action( 'admin_init', array( $this, 'migrate' ) );
+	}
+
+	/**
+	 * Migrates the DB.
+	 *
+	 * @since 1.128.0
+	 */
+	public function migrate() {
+		$db_version = $this->options->get( self::DB_VERSION_OPTION );
+
+		if ( ! $db_version || version_compare( $db_version, self::DB_VERSION, '<' ) ) {
+			$this->migrate_analytics_conversion_id_setting();
+			$this->activate_ads_module();
+
+			$this->options->set( self::DB_VERSION_OPTION, self::DB_VERSION );
+		}
+	}
+
+	/**
+	 * Migrates the Ads Conversion ID to the new Ads module.
+	 *
+	 * @since 1.128.0
+	 */
+	protected function migrate_analytics_conversion_id_setting() {
+		if ( ! $this->analytics_settings->has() ) {
+			return;
+		}
+
+		$analytics_settings = $this->options->get( self::ANALYTICS_OPTIONS );
+
+		if ( empty( $analytics_settings ) || ! array_key_exists( 'adsConversionID', $analytics_settings ) || empty( $analytics_settings['adsConversionID'] ) ) {
+			return;
+		}
+
+		$ads_settings = $this->options->get( self::ADS_OPTIONS );
+
+		if ( array_key_exists( 'conversionID', $ads_settings ) && ! empty( $ads_settings['conversionID'] ) ) {
+			// If there is already an adsConversionID set in the Ads module, do not overwrite it, remove it from the Analytics module.
+			$this->options->set( self::ANALYTICS_OPTIONS, array_merge( $analytics_settings, array( 'adsConversionID' => '' ) ) );
+
+			return;
+		}
+
+		$new_ads_settings = array_merge(
+			$ads_settings,
+			array(
+				'conversionID' => $analytics_settings['adsConversionID'],
+			)
+		);
+		$this->options->set( self::ADS_OPTIONS, array_merge( $ads_settings, $new_ads_settings ) );
+
+		$new_analytics_settings = array_merge(
+			$analytics_settings,
+			array(
+				'adsConversionID'             => '',
+				'adsConversionIDMigratedAtMs' => time() * 1000,
+			)
+		);
+		$this->options->set( self::ANALYTICS_OPTIONS, array_merge( $analytics_settings, $new_analytics_settings ) );
```

Comment:
The same here, if we want to remove a few keys from settings, then we should use `unset`:

```php
unset( $analytics_settings['adsConversionID'], $analytics_settings['adsConversionIDMigratedAtMs'] );
$this->options->set( self::ANALYTICS_OPTIONS, $analytics_settings );
```

Reply:
Good points here, and above, I've simplified several options updates.